packages/nitro-protocol/contracts/ninja-nitro/readme.md

```diff
@@ -0,0 +1,97 @@
+
+### Ninja-Nitro
+
+Ninja-Nitro, by contrast has 1 x `AdjudicatorFactory` contract, 1 x `SingleChannelAdjudicator` master contract, and k x `Proxy` contracts, one for each of the k channels in the network that resolves on chain. The factory and master contracts are "infastructure" and deployed once (most likely by the protocol team).
+
+The `AdjudicatorFactory` API is
+
+- `createChannel`
+- utility methods that batch `createChannel` with the master contract API.
+
+The `createChannel` method will deploy a new `Proxy` for each channel as necessary (namely, if it needs to be directly defunded). Such `Proxy` contracts are lightweight (cheap to deploy), have easily predictable addresses, and point to the `SingleChannelAdjudicator` master contract. They therefore have the same API:
+
+- The ForceMove API
+- `transfer`
+- `claim`
+- utility methods that batch up this API
+
+Note however, that the Proxy contracts each have their own storage. See https://eips.ethereum.org/EIPS/eip-1167.
+
+## Similarities and differences
+
+### Holdings
+
+The Mono-Nitro system has m x `holdings` mappings, one in each AssetHolder. This mapping is `public`, so has a public getter function.
+Ninja-Nitro has no such mappings: but a public `holdings` function remains. This functions simply reads the balance of the current contract (or another contract) from either i) the Ethereum accounting system or ii) a given Token contract.
+
+### Deposits
+
+In Mono-Nitro, depositing requires calling a method on each `AssetHolder`.
+In Ninja-Nitro, depositing can be achieved simply by sending funds or tokens to the address of the `Proxy`, and this can even be done before the `Proxy` is created/deployed.
+
+### Adjudicator status
+
+Ninja-Nitro was implemented by modifying a copy of Mono-Nitro. This means that certain quirks remain in the implementation which would probably not exist had it been written from scratch. These quirks may or may not be removed in future.
+
+One such quirk is as follows. In Mono-Nitro, the hash of the status of a channel (which includes the full outcome, the address of any challenger, the time when the channel will finalize, etc) is stored in a mapping, keyed by the `channelId`.
+
+In Ninja-Nitro, exactly the same is true: only each proxy has its own mapping. The reason that this is quirky is that each of these mappings is only ever going to have a single entry, because each proxy is in a 1-1 relation to a `channelId`. The reason that this quirk is not actually very troublesome are:
+
+- the gas cost are not dramatically higher than storing a single hash without using a mapping (it may in fact cost precisely the same amount). See https://docs.soliditylang.org/en/v0.8.3/types.html?highlight=mapping#mapping-types
```

Comment:
This does not feel possible to me. If a mapping contains a `uint` number , then 32 bytes would be required to store the number. If those 32 bytes are the only thing stored, how can the bookkeeping happen?

Are you sure that there's no potential gas savings from switching from a `mapping bytes32 -> bytes32` state variable to a `bytes32` state variable?

https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays suggests a mapping has 32 bytes of overhead in addition to the storage that their items contain. Thus, I would predict that a `mapping bytes32 -> bytes32` would require at least two storage slots to store one `bytes32` key/value pair.

Reply:
> This does not feel possible to me. If a mapping contains a `uint` number , then 32 bytes would be required to store the number. If those 32 bytes are the only thing stored, how can the bookkeeping happen?
> 

Clearly some extra information must be stored for bookkeeping; but the question here is about whether there is more bookkeeping involved in storing something inside a mapping rather than as a simple state variable. In each case the location of the stored data in storage must be recorded. From what I can tell from the docs, determining the location of a mapping key involves computing the hash of that key concatenated with the mapping location itself. So I think there is going to be slightly elevated cost to using a mapping.

> Are you sure that there's no potential gas savings from switching from a `mapping bytes32 -> bytes32` state variable to a `bytes32` state variable?

I suspect there will not be significant savings, but I can't say I am sure there will be no savings at all. By significant I mean it won't be 2 storage slots versus 1 storage slot (20K difference in gas). 

> 
> https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#mappings-and-dynamic-arrays suggests a mapping has 32 bytes of overhead in addition to the storage that their items contain. Thus, I would predict that a `mapping bytes32 -> bytes32` would require at least two storage slots to store one `bytes32` key/value pair.

I think the extra slot you are referring to, which records the location of the mapping in storage, is reserved but not written to, so doesn't incur gas costs. 

I'll knock together a quick investigation to see if we can check this.

Reply:
Investigation [here](https://github.com/statechannels/statechannels/compare/mapping-investigation?expand=1) implies the costs are

Reply:
> I think the extra slot you are referring to, which records the location of the mapping in storage, is reserved but not written to, so doesn't incur gas costs.

Ah, thanks for clarifying. I hadn't considered that you could reserve the storage slot without writing to it.

Thanks for investigating this!